lib/vnet/app_resolver.go

```diff
@@ -282,49 +283,88 @@ func (r *TCPAppResolver) resolveTCPHandlerForCluster(
 }
 
 type tcpAppHandler struct {
-	profileName     string
-	leafClusterName string
-	lp              *alpnproxy.LocalProxy
+	log              *slog.Logger
+	appProvider      AppProvider
+	clock            clockwork.Clock
+	profileName      string
+	leafClusterName  string
+	app              types.Application
+	portToLocalProxy map[uint16]*alpnproxy.LocalProxy
+	// mu guards access to portToLocalProxy.
+	mu sync.Mutex
 }
 
 func (r *TCPAppResolver) newTCPAppHandler(
 	ctx context.Context,
+	log *slog.Logger,
 	profileName string,
 	leafClusterName string,
 	app types.Application,
 ) (*tcpAppHandler, error) {
-	dialOpts, err := r.appProvider.GetDialOptions(ctx, profileName)
+	return &tcpAppHandler{
+		appProvider:      r.appProvider,
+		clock:            r.clock,
+		profileName:      profileName,
+		leafClusterName:  leafClusterName,
+		app:              app,
+		portToLocalProxy: make(map[uint16]*alpnproxy.LocalProxy),
+		log:              log.With(teleport.ComponentKey, "VNet.AppHandler"),
+	}, nil
+}
+
+// getOrInitializeLocalProxy returns a separate local proxy for each port for multi-port apps. For
+// single-port apps, it returns the same local proxy no matter the port.
+func (h *tcpAppHandler) getOrInitializeLocalProxy(ctx context.Context, localPort uint16) (*alpnproxy.LocalProxy, error) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	// Connections to single-port apps need to go through a local proxy that has a cert with TargetPort
+	// set to 0. This ensures that the old behavior is kept for such apps, where the client can dial
+	// the public address of an app on any port and be routed to the port from the URI.
+	//
+	// https://github.com/gravitational/teleport/blob/master/rfd/0182-multi-port-tcp-app-access.md#vnet-with-single-port-apps
+	if len(h.app.GetTCPPorts()) == 0 {
+		localPort = 0
+	}
```

Comment:
If it is a multi-port app, I think we need to check that the localPort matches one of the tcp ports in the app spec. The RFD does mention that the VNet client should do that. I think it would prevent creating too many cert requests if something port-scans the local IP address

I tried a non-matching port locally and I noticed it's actually able to get a cert for the app with that port and dial all the way to the app server. I think the auth server should be rejecting non-matching ports in cert requests too

Reply:
actually, it's probably not necessary to do the port match check on the auth server, it looks like we don't really do access checks for apps in generateUserCert, I don't think we ever even fetch the app spec. I think we should check it in VNet though

Reply:
I do plan to check the port in one of the upcoming PRs and surface it in the Electron app.